File: apple/HDHROpenKit/Sources/HDHROpenKit/Utilities/GuideGridMath.swift
import CoreGraphics
import Foundation

/// Pure layout math mirroring the web guide grid (frontend/src/lib/components/details/HDHomeRunGuideGrid.svelte)
/// and Android client (android/app/src/main/kotlin/org/hdhropen/app/ui/screens/guide/GuideGridView.kt).
public enum GuideGridMath {
    public static let pxPerSecond: CGFloat = 4.0 / 60.0
    public static let minCellWidth: CGFloat = 90
    public static let hourSeconds: TimeInterval = 3600
    public static let daySeconds: TimeInterval = 86400

    public static func windowBounds(
        nowSeconds: TimeInterval,
        fullGuide: [HDHomeRunFullGuideChannel]
    ) -> (start: TimeInterval, end: TimeInterval) {
        var minStart = nowSeconds - 2 * hourSeconds
        var maxEnd = nowSeconds + 4 * hourSeconds
        let earliestAllowed = nowSeconds - 6 * hourSeconds
        let latestAllowed = nowSeconds + 48 * hourSeconds

        for entry in fullGuide {
            for airing in entry.airings {
                if let start = airing.start {
                    minStart = min(minStart, start)
                }
                if let end = airing.end {
                    maxEnd = max(maxEnd, end)
                }
            }
        }

        minStart = max(minStart, earliestAllowed)
        maxEnd = min(maxEnd, latestAllowed)
        let start = (minStart / 1800).rounded(.down) * 1800
        return (start, maxEnd)
    }

    /// Computes the horizontal scroll offset in points so "now" is offset by `leadingPadding`
    /// from the leading edge of the timeline track. Clamped to >= 0.
    /// Mirrors web `Math.max(nowLeft - 60, 0)` and Android `(nowLeft - 60.dp).coerceAtLeast(0)`.
    public static func targetScrollOffset(
        nowSeconds: TimeInterval,
        windowStart: TimeInterval,
        leadingPadding: CGFloat = 60
    ) -> CGFloat {
        let nowLeft = CGFloat(nowSeconds - windowStart) * pxPerSecond
        return max(nowLeft - leadingPadding, 0)
    }

    public struct CellLayout: Identifiable, Sendable {
        public var id: String {
            airing.id
        }

        public let airing: HDHomeRunGuideEntry
        public let left: CGFloat
        public let width: CGFloat

        public init(airing: HDHomeRunGuideEntry, left: CGFloat, width: CGFloat) {
            self.airing = airing
            self.left = left
            self.width = width
        }
    }

    public static func cellLayouts(
        airings: [HDHomeRunGuideEntry],
        windowStart: TimeInterval,
        windowEnd: TimeInterval
    ) -> [CellLayout] {
        var layouts: [CellLayout] = []
        for airing in airings {
            guard let airingStart = airing.start, let airingEnd = airing.end else { continue }
            let start = max(airingStart, windowStart)
            let end = min(airingEnd, windowEnd)
            guard end > start else { continue }
            let left = CGFloat(start - windowStart) * pxPerSecond
            let width = max(CGFloat(end - start) * pxPerSecond, minCellWidth)
            layouts.append(CellLayout(airing: airing, left: left, width: width))
        }
        return layouts
    }

    public struct HourMark: Identifiable, Sendable {
        public var id: TimeInterval {
            seconds
        }

        public let seconds: TimeInterval
        public let left: CGFloat
        public let label: String

        public init(seconds: TimeInterval, left: CGFloat, label: String) {
            self.seconds = seconds
            self.left = left
            self.label = label
        }
    }

    public struct DayMark: Identifiable, Sendable {
        public var id: TimeInterval {
            start
        }

        public let start: TimeInterval
        public let left: CGFloat
        public let width: CGFloat
        public let label: String

        public init(start: TimeInterval, left: CGFloat, width: CGFloat, label: String) {
            self.start = start
            self.left = left
            self.width = width
            self.label = label
        }
    }

    private static let hourFormatter: DateFormatter = {
        let f = DateFormatter()
        f.dateFormat = "h a"
        return f
    }()

    private static let weekdayFormatter: DateFormatter = {
        let f = DateFormatter()
        f.dateFormat = "EEEE"
        return f
    }()

    public static func hourMarks(windowStart: TimeInterval, windowEnd: TimeInterval) -> [HourMark] {
        var marks: [HourMark] = []
        let calendar = Calendar.current
        let startDate = Date(timeIntervalSince1970: windowStart)
        let endDate = Date(timeIntervalSince1970: windowEnd)
        guard var cursor = calendar.dateInterval(of: .hour, for: startDate)?.start else { return marks }

        while cursor < endDate {
            let seconds = cursor.timeIntervalSince1970
            if seconds >= windowStart {
                let left = CGFloat(seconds - windowStart) * pxPerSecond
                marks.append(HourMark(seconds: seconds, left: left, label: hourFormatter.string(from: cursor)))
            }
            guard let next = calendar.date(byAdding: .hour, value: 1, to: cursor) else { break }
            cursor = next
        }
        return marks
    }

    public static func dayMarks(windowStart: TimeInterval, windowEnd: TimeInterval) -> [DayMark] {
        var marks: [DayMark] = []
        let calendar = Calendar.current
        let endDate = Date(timeIntervalSince1970: windowEnd)
        guard var dayStart = calendar.dateInterval(of: .day, for: Date(timeIntervalSince1970: windowStart))?.start else {
            return marks
        }
        let today = calendar.startOfDay(for: Date())

        while dayStart < endDate {
            guard let dayEnd = calendar.date(byAdding: .day, value: 1, to: dayStart) else { break }
            let segStart = max(dayStart.timeIntervalSince1970, windowStart)
            let segEnd = min(dayEnd.timeIntervalSince1970, windowEnd)
            if segEnd > segStart {
                let left = CGFloat(segStart - windowStart) * pxPerSecond
                let width = CGFloat(segEnd - segStart) * pxPerSecond
                let label: String = if calendar.isDate(dayStart, inSameDayAs: today) {
                    "Today"
                } else if let tomorrow = calendar.date(byAdding: .day, value: 1, to: today),
                          calendar.isDate(dayStart, inSameDayAs: tomorrow)
                {
                    "Tomorrow"
                } else {
                    weekdayFormatter.string(from: dayStart)
                }
                marks.append(DayMark(start: dayStart.timeIntervalSince1970, left: left, width: width, label: label))
            }
            dayStart = dayEnd
        }
        return marks
    }
}

File: android/core/src/main/kotlin/org/hdhropen/kit/utilities/GuideGridMath.kt
package org.hdhropen.kit.utilities

import org.hdhropen.kit.models.HDHomeRunFullGuideChannel
import org.hdhropen.kit.models.HDHomeRunGuideEntry
import kotlin.math.max
import kotlin.math.min

object GuideGridMath {
    const val DP_PER_SECOND = 4.0f / 60.0f // 4dp per minute
    const val MIN_CELL_WIDTH = 90.0f
    const val HOUR_SECONDS = 3600.0
    const val DAY_SECONDS = 86400.0

    fun windowBounds(
        nowSeconds: Double,
        fullGuide: List<HDHomeRunFullGuideChannel>
    ): Pair<Double, Double> {
        var minStart = nowSeconds - 2 * HOUR_SECONDS
        var maxEnd = nowSeconds + 4 * HOUR_SECONDS
        val earliestAllowed = nowSeconds - 6 * HOUR_SECONDS

        for (entry in fullGuide) {
            for (airing in entry.airings) {
                airing.start?.let { minStart = min(minStart, it) }
                airing.end?.let { maxEnd = max(maxEnd, it) }
            }
        }

        minStart = max(minStart, earliestAllowed)
        val start = kotlin.math.floor(minStart / 1800.0) * 1800.0
        return Pair(start, maxEnd)
    }

    data class CellLayout(
        val airing: HDHomeRunGuideEntry,
        val leftDp: Float,
        val widthDp: Float
    )

    fun cellLayouts(
        airings: List<HDHomeRunGuideEntry>,
        windowStart: Double,
        windowEnd: Double
    ): List<CellLayout> {
        val layouts = mutableListOf<CellLayout>()
        for (airing in airings) {
            val airingStart = airing.start ?: continue
            val airingEnd = airing.end ?: continue
            val start = max(airingStart, windowStart)
            val end = min(airingEnd, windowEnd)
            if (end <= start) continue

            val left = ((start - windowStart) * DP_PER_SECOND).toFloat()
            val calculatedWidth = ((end - start) * DP_PER_SECOND).toFloat()
            val width = maxOf(calculatedWidth, MIN_CELL_WIDTH)

            layouts.add(CellLayout(airing, left, width))
        }
        return layouts
    }

    data class HourMark(val seconds: Double, val leftDp: Float, val label: String)
    data class DayMark(val start: Double, val leftDp: Float, val widthDp: Float, val label: String)

    fun hourMarks(windowStart: Double, windowEnd: Double): List<HourMark> {
        val marks = mutableListOf<HourMark>()
        var cursor = (kotlin.math.floor(windowStart / 3600.0) * 3600.0)
        while (cursor < windowEnd) {
            if (cursor >= windowStart) {
                val left = ((cursor - windowStart) * DP_PER_SECOND).toFloat()
                marks.add(HourMark(cursor, left, TimeFormatting.formatHour(cursor)))
            }
            cursor += 3600.0
        }
        return marks
    }

    fun dayMarks(windowStart: Double, windowEnd: Double): List<DayMark> {
        val marks = mutableListOf<DayMark>()
        var cursor = (kotlin.math.floor(windowStart / 86400.0) * 86400.0)
        while (cursor < windowEnd) {
            val dayEnd = cursor + 86400.0
            val segStart = max(cursor, windowStart)
            val segEnd = min(dayEnd, windowEnd)
            if (segEnd > segStart) {
                val left = ((segStart - windowStart) * DP_PER_SECOND).toFloat()
                val width = ((segEnd - segStart) * DP_PER_SECOND).toFloat()
                marks.add(DayMark(cursor, left, width, TimeFormatting.formatDayLabel(segStart)))
            }
            cursor = dayEnd
        }
        return marks
    }
}
